example_buffer.py

```diff
@@ -0,0 +1,194 @@
+
+    def update(self, new_games, samples_per_game=4):
+        """
+        new_games is list of .tfrecord.zz files of new games
+        """
+        new_games.sort(key=lambda f: os.path.basename(f))
+        first_new_game = None
+        for i, game in enumerate(tqdm(new_games)):
+            t = file_timestamp(game)
+            if t <= self.examples[-1][0]:
+                continue
+            elif first_new_game is None:
+                first_new_game = i
+                print("Found {}/{} new games".format(len(new_games)-i, len(new_games)))
+
+            choices = [(t, ex) for ex in pick_examples_from_tfrecord(
+                game, samples_per_game)]
+            if self.count > self.max_size:
+                self.examples = self.examples[samples_per_game:]
+            self.examples.extend(choices)
+
+    def flush(self, path):
+        with timer("Writing examples to " + path):
+            preprocessing.write_tf_examples(
+                path, [ex[1] for ex in self.examples])
+        del self.examples
+        self.examples = []
+
+    @property
+    def count(self):
+        return len(self.examples)
+
+    def __str__(self):
+        return "ExampleBuffer: {} positions sampled from {} to {}".format(
+            self.count,
+            _ts_to_str(self.examples[0][0]),
+            _ts_to_str(self.examples[-1][0]))
+
+
+def files_for_model(model):
+    return tf.gfile.Glob(os.path.join(LOCAL_DIR, model[1], '*.zz'))
+
+
+def smart_rsync(from_model_num=0, source_dir=rl_loop.SELFPLAY_DIR, dest_dir=LOCAL_DIR):
+    from_model_num = 0 if from_model_num < 0 else from_model_num
+    models = [m for m in rl_loop.get_models() if m[0] >= from_model_num]
+    for m in models:
+        _rsync_dir(os.path.join(
+            source_dir, m[1]), os.path.join(dest_dir, m[1]))
+
+
+def _rsync_dir(source_dir, dest_dir):
+    _ensure_dir_exists(dest_dir)
+    subprocess.call(['gsutil', '-m', 'rsync', source_dir, dest_dir],
+                    stderr=open('.rsync_log', 'ab'))
+
+
+def loop(bufsize=dual_net.EXAMPLES_PER_GENERATION,
+         write_dir=rl_loop.TRAINING_CHUNK_DIR,
+         model_window=100,
+         threads=8,
+         skip_first_rsync=False):
+    buf = ExampleBuffer(bufsize)
+
+    while True:
+        models = rl_loop.get_models()[-model_window:]
+        if not skip_first_rsync:
+            with timer("Rsync"):
+                smart_rsync(models[-1][0] - 6)
+        files = list(tqdm(map(files_for_model, models), total=len(models)))
+        buf.parallel_fill(list(itertools.chain(*files)), threads=threads)
+
+        print("Filled buffer, watching for new games")
+        while rl_loop.get_latest_model()[0] == models[-1][0]:
+            with timer("Rsync"):
+                smart_rsync(models[-1][0] - 2)
+            new_files = list(
+                tqdm(map(files_for_model, models[-2:]), total=len(models)))
+            buf.update(list(itertools.chain(*new_files)))
+            print(self)
```

Comment:
probably have some small sleep here, in practice we know rsync is slow but it's nice to limit things.